lms/djangoapps/courseware/courses.py

```diff
@@ -183,20 +236,22 @@ def check_course_access(course, user, action, check_if_enrolled=False, check_sur
                 dashboard_url=reverse('dashboard'),
             ), access_response)
 
-        # Deliberately return a non-specific error message to avoid
-        # leaking info about access control settings
-        raise CoursewareAccessException(access_response)
+        # Redirect if the user is not enrolled and must be to see content
+        if isinstance(access_response, EnrollmentRequiredAccessError):
+            raise CourseAccessRedirect(reverse('about_course', args=[six.text_type(course.id)]))
 
-    if check_if_enrolled:
```

Comment:
You can see in the new code that this enrollment check has been folded into check_course_access, along with the survey below.  Then, in check_course_access_with_redirect, we can treat those two conditions like any of the others above.